### PR TITLE
Introduce more types and add type conversion

### LINKED ISSUE #67: Casting from int to float and float to int
We should be able to cast floating point values from integer values . We do not have equivalent C code and its LLVM IR issue.
for ex : 
```
(declare a int)
(set a 10)
(declare b float)
(set b 10.0)
(declare c float)
(set c (fadd a b ))
```
We cannot add 'a' and 'b' as both of them are in different formats. We should be able to change the format of 'a' to float before adding it.

We should be able to do something similar to the below : 

```
(c-lisp

    (define ((fprint float) (n  float) ))

    (define ((main void))
    (declare a int)
    (declare b float)
    (declare c float)
    (set a 10)
    (set b 10.0)
    (declare a float)
    (set c (fadd a b))
    (call fprint c)

    )
)
```
The output when the above is executed executed : 

```
guile ../../utils/sexp-json.scm < test.sexp  | python ../../c-lisp.py | python ../../brilisp.py | python ../../llvm.py | bash ../brilisp/run.sh {args}
Error in statement: ['declare', 'a', 'float']
Traceback (most recent call last):
  File "/home/aadithya.bhat/c_lisp/bril/bril-txt/llama.lisp/src/backend/tests/c-lisp/../../c-lisp.py", line 473, in <module>
    json.dump(brilisp_code_generator.c_lisp(json.loads(sys.stdin.read())), sys.stdout)
  File "/home/aadithya.bhat/c_lisp/bril/bril-txt/llama.lisp/src/backend/tests/c-lisp/../../c-lisp.py", line 59, in c_lisp
    return ["brilisp"] + [self.gen_function(fn) for fn in prog[1:]]
  File "/home/aadithya.bhat/c_lisp/bril/bril-txt/llama.lisp/src/backend/tests/c-lisp/../../c-lisp.py", line 59, in <listcomp>
    return ["brilisp"] + [self.gen_function(fn) for fn in prog[1:]]
  File "/home/aadithya.bhat/c_lisp/bril/bril-txt/llama.lisp/src/backend/tests/c-lisp/../../c-lisp.py", line 125, in gen_function
    *self.gen_compound_stmt(
  File "/home/aadithya.bhat/c_lisp/bril/bril-txt/llama.lisp/src/backend/tests/c-lisp/../../c-lisp.py", line 303, in gen_compound_stmt
    instr_list += self.gen_stmt(s)
  File "/home/aadithya.bhat/c_lisp/bril/bril-txt/llama.lisp/src/backend/tests/c-lisp/../../c-lisp.py", line 163, in gen_stmt
    raise e
  File "/home/aadithya.bhat/c_lisp/bril/bril-txt/llama.lisp/src/backend/tests/c-lisp/../../c-lisp.py", line 150, in gen_stmt
    return self.gen_decl_stmt(stmt)
  File "/home/aadithya.bhat/c_lisp/bril/bril-txt/llama.lisp/src/backend/tests/c-lisp/../../c-lisp.py", line 271, in gen_decl_stmt
    raise CodegenError(f"Re-declaration of variable {name}")
__main__.CodegenError: Re-declaration of variable a
Traceback (most recent call last):
  File "/home/aadithya.bhat/c_lisp/bril/bril-txt/llama.lisp/src/backend/tests/c-lisp/../../brilisp.py", line 174, in <module>
    main()
  File "/home/aadithya.bhat/c_lisp/bril/bril-txt/llama.lisp/src/backend/tests/c-lisp/../../brilisp.py", line 169, in main
    expr = json.load(sys.stdin)
  File "/gnu/store/kgxbvcr9ks2k8lbm8fla7zvp63dbagrc-python-3.10.7/lib/python3.10/json/__init__.py", line 293, in load
    return loads(fp.read(),
  File "/gnu/store/kgxbvcr9ks2k8lbm8fla7zvp63dbagrc-python-3.10.7/lib/python3.10/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "/gnu/store/kgxbvcr9ks2k8lbm8fla7zvp63dbagrc-python-3.10.7/lib/python3.10/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/gnu/store/kgxbvcr9ks2k8lbm8fla7zvp63dbagrc-python-3.10.7/lib/python3.10/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
Traceback (most recent call last):
  File "/home/aadithya.bhat/c_lisp/bril/bril-txt/llama.lisp/src/backend/tests/c-lisp/../../llvm.py", line 359, in <module>
    main()
  File "/home/aadithya.bhat/c_lisp/bril/bril-txt/llama.lisp/src/backend/tests/c-lisp/../../llvm.py", line 352, in main
    bril_prog = munch.munchify(json.load(sys.stdin))
  File "/gnu/store/kgxbvcr9ks2k8lbm8fla7zvp63dbagrc-python-3.10.7/lib/python3.10/json/__init__.py", line 293, in load
    return loads(fp.read(),
  File "/gnu/store/kgxbvcr9ks2k8lbm8fla7zvp63dbagrc-python-3.10.7/lib/python3.10/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "/gnu/store/kgxbvcr9ks2k8lbm8fla7zvp63dbagrc-python-3.10.7/lib/python3.10/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/gnu/store/kgxbvcr9ks2k8lbm8fla7zvp63dbagrc-python-3.10.7/lib/python3.10/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
/usr/bin/ld: /gnu/store/ln6hxqjvz6m9gdd9s97pivlqck7hzs99-glibc-2.35/lib/Scrt1.o: in function `_start':
(.text+0x17): undefined reference to `main'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
../brilisp/run.sh: line 9: /tmp/tmp.InjTJQHqDK.out: No such file or directory
rm: cannot remove '/tmp/tmp.InjTJQHqDK.out': No such file or directory
```

### PR DESCRIPTION
**Pull request by adi-lb-phoenix**
I have implemented 'Int to Float' and 'Float to Int' explicit type casting in Brilisp which fixes #67.
The syntax for it is `(set (f_n float) (sitofp n))` in case integer to float and ` (set (f_n int) (fptosi n))` in case of float to integer explicit conversions.

### CODE REVIEW

File: src/backend/tests/brilisp/castop_fptoui.sexp [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    (define ((print int32) (n int32)))
+
+    (define ((cas_t int32) (n double))
+        (set (f_n int32) (fptoui n int32))
+        (ret f_n))
+
+    (define ((main void))
+        (set (a double) (const 123.0))
+        (set (b int32) (call cas_t a))
+        (set (tmp int32) (call print b))
```
Comment (chsasank) [inline src/backend/tests/brilisp/castop_fptoui.sexp:14]: can you cast negative number and see what happens here?

File: src/backend/tests/brilisp/castop_fprunc.sexp
Comment (chsasank) [inline src/backend/tests/brilisp/castop_fprunc.sexp:1]: duplicate. fptrunc is already there

File: src/backend/tests/brilisp/castop_inttoptr_ptrtoint.sexp [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    (define ((cast_inttopointer (ptr int)) (n int))
+        (set (f_n (ptr int))(inttoptr n (ptr int)))
+        (ret f_n))
+    
+    (define ((cast_pointertoint int) (n (ptr int)))
+        (set (return_int int) (ptrtoint n int))
+        (ret return_int))
+
+    (define ((main void))
+        (set (a int) (const 255))
```
Comment (chsasank) [inline src/backend/tests/brilisp/castop_inttoptr_ptrtoint.sexp:13]: can you add comments what you did here

File: src/backend/tests/brilisp/castop_sitofp_F_to_F.sexp [hunk 3]
```diff
@@ -0,0 +1,14 @@
+(brilisp
+    (define ((fprint float) (n float)))
+
+    (define ((add5 float) (n float))
```
Comment (chsasank) [inline src/backend/tests/brilisp/castop_sitofp_F_to_F.sexp:4]: add a test with negative numbers as well

File: src/backend/tests/brilisp/castop_uitofp.sexp [hunk 4]
```diff
@@ -0,0 +1,21 @@
+(brilisp
+    (define ((fprint float) (n float)))
+
+    (define ((dprint double) (n double)))
+
+    (define ((cas_t float) (n int32))
```
Comment (chsasank) [inline src/backend/tests/brilisp/castop_uitofp.sexp:6]: can you ditch functions in all the tests and directly put everything in tests please?

File: src/backend/tests/brilisp/int_to_float.sexp
Comment (chsasank) [inline src/backend/tests/brilisp/int_to_float.sexp:1]: aren't these duplicates?

File: src/backend/tests/brilisp/castop_fptoui.sexp (hunk 1 above)
Comment (adi-lb-phoenix) [reply src/backend/tests/brilisp/castop_fptoui.sexp:14]: Tested it . Its fp to unsigned integer. For the below code.
```
(set (c double) (const -123.00000))
        (set (d int32) (call cas_t c))
        (set (tmp_1 int32) (call print d))
        (ret )))
```
For every execution `fptoui` on a negative integer has been giving the results of size similar to number of digits in 484409152. The value can be both positive and negative.

File: src/backend/tests/brilisp/castop_sext.sexp [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+    (define ((print int16) (n int16)))
+
+    (define ((cas_t int16) (n int8))
+        (set (f_n int16) (sext n int8))
+        (ret f_n))
+
+    (define ((main void))
+        (set (a int8) (const -1))
+        (set (b int16) (call cas_t a))
+        (set (tmp int16) (call print b))
```
[the remark this answers is not in the document]
Comment (adi-lb-phoenix) [reply src/backend/tests/brilisp/castop_sext.sexp:12]:
```
(brilisp
    (define ((print int32) (n int32)))

    (define ((cas_t int32) (n int8))
        (set (f_n int16) (sext n int8))
        (set (f_n_1 int32) (sext f_n int32))
        (ret f_n_1))

    (define ((main void))
        (set (a int8) (const -1))
        (set (b int32) (call cas_t a))
        (set (tmp int32) (call print b))
        (ret)))

```
The above code will give you -1.  So here it is getting converted to i16 . When we call print in the above which expects a 32bit integer. But we send only 16bit integer as a result printf interprets first 16 bits to contain 0 , the next 16 bits contain to 1. The decimal value of 16 bits containing one is 65535 . Hence when we get the 65535. If we want -1 as the output. We should explicitly convert i16 to i32 using `sext` after which when printed we get the result -1.

**Comment by chsasank** [comment]
Awesome. Merging.
